include/deal.II/optimization/line_minimization.h

```diff
@@ -0,0 +1,501 @@
+
+  /**
+   * Perform cubic fit if gradients have different sign, otherwise fallback
+   * to quadratic.
+   */
+  template <typename NumberType>
+  NumberType
+  poly_fit(const NumberType                        x_low,
+           const NumberType                        f_low,
+           const NumberType                        g_low,
+           const NumberType                        x_hi,
+           const NumberType                        f_hi,
+           const NumberType                        g_hi,
+           const NumberType                        x_rec,
+           const NumberType                        f_rec,
+           const unsigned int                      i,
+           const std::pair<NumberType, NumberType> bounds);
+
+  /**
+   * Perform a line search in $(0,max]$ with strong Wolfe conditions
+   * \f[
+   * f(\alpha) \le f(0) + \alpha \mu f'(0) \\
+   * |f'(\alpha)| \le \eta |f'(0)|
+   * \f]
+   * using one dimension
+   * functions @p func and a function @p interpolate to choose a new point
+   * from the interval based on the function values and derivatives at its ends.
+   * @p a1 is a trial estimate of the step.
+   *
+   * The function implements Algorithms
+   * 2.6.2 and 2.6.4 on pages 34-35 in Fletcher, 2013, Practical methods of
+   * optimization. that are minor variations of the Algorithm 3.5 and 3.6 on
+   * pages 60-61 in Nocedal and Wright, Numerical optimization.
+   * It consists of bracketing phase and the zoom phase, where @p interpolate is used.
+   *
+   * The function returns the step size and the number of times function @p func was called.
+   */
+  template <typename NumberType>
+  std::pair<NumberType, unsigned int>
+  line_search(
+    const std::function<void(const NumberType &x, NumberType &f, NumberType &g)>
+      &              func,
+    const NumberType f0,
+    const NumberType g0,
+    const std::function<
+      NumberType(const NumberType                        x_low,
+                 const NumberType                        f_low,
+                 const NumberType                        g_low,
+                 const NumberType                        x_hi,
+                 const NumberType                        f_hi,
+                 const NumberType                        g_hi,
+                 const NumberType                        x_rec,
+                 const NumberType                        g_rec,
+                 const unsigned int                      i,
+                 const std::pair<NumberType, NumberType> bounds)> &interpolate,
+    const NumberType                                               a1,
+    const NumberType                                               eta = 0.9,
+    const NumberType                                               mu  = 0.01,
+    const NumberType   a_max           = std::numeric_limits<NumberType>::max(),
+    const bool         debug_output    = false,
+    const unsigned int max_evaluations = 20);
```

Comment:
there are a lot of function arguments here whose meaning is not obvious, including the ones to `interpolate` and `func`. Can you document them individually?

Reply:
And their descriptions need the proper punctuation and capitalisation as well.